mypy/types.py

```diff
@@ -446,13 +447,63 @@ def deserialize(cls, data: JsonDict) -> 'TypeVarType':
         )
 
 
+class ParamSpecFlavor:
+    # Simple ParamSpec reference such as "P"
+    BARE: Final = 0
+    # P.args
+    ARGS: Final = 1
+    # P.kwargs
+    KWARGS: Final = 2
+
+
 class ParamSpecType(TypeVarLikeType):
-    """Definition of a single ParamSpec variable."""
+    """Type that refers to a ParamSpec.
 
-    __slots__ = ()
+    A ParamSpec is a type variable that represents the parameter
+    types, names and kinds of a callable (i.e., the signature without
+    the return type).
 
-    def __repr__(self) -> str:
-        return self.name
+    This can be one of these forms
+     * P (ParamSpecFlavor.BARE)
+     * P.args (ParamSpecFlavor.ARGS)
+     * P.kwargs (ParamSpecFLavor.KWARGS)
+    """
+
+    __slots__ = ('flavor',)
+
+    flavor: int
+
+    def __init__(
+         self, name: str, fullname: str, id: Union[TypeVarId, int], flavor: int,
+         upper_bound: Type, *, line: int = -1, column: int = -1
+    ) -> None:
+        super().__init__(name, fullname, id, upper_bound, line=line, column=column)
+        self.flavor = flavor
+
+    @staticmethod
+    def new_unification_variable(old: 'ParamSpecType') -> 'ParamSpecType':
+        new_id = TypeVarId.new(meta_level=1)
+        return ParamSpecType(old.name, old.fullname, new_id, old.flavor, old.upper_bound,
+                             line=old.line, column=old.column)
+
+    def accept(self, visitor: 'TypeVisitor[T]') -> T:
+        return visitor.visit_param_spec(self)
+
+    def name_with_suffix(self) -> str:
+        n = self.name
+        if self.flavor == ParamSpecFlavor.ARGS:
+            return f'{n}.args'
+        elif self.flavor == ParamSpecFlavor.KWARGS:
+            return f'{n}.kwargs'
+        return n
+
+    def __hash__(self) -> int:
+        return hash((self.id, self.flavor))
+
+    def __eq__(self, other: object) -> bool:
+        if not isinstance(other, ParamSpecType):
+            return NotImplemented
+        return self.id == other.id and self.flavor == other.flavor
```

Comment:
Why `upper_bound` is missing here and in various other places like `is_same_type()`? Add a comment about this.

Reply:
It's not user-configurable and always depends on flavor (actually currently it's always `object`). Added some comments.